framework/resource/resource.go

```diff
@@ -149,6 +152,34 @@ func (r *Resource) Destroy(args ...interface{}) error {
 	return result.Err()
 }
 
+// DeclaredResource converts a resource to a DeclaredResource protobuf, which
+// can be used in a component.DeclaredResourcesResp.
+func (r *Resource) DeclaredResource() (*pb.DeclaredResource, error) {
+	stateJson, err := json.Marshal(r.State())
+	if err != nil {
+		return nil, fmt.Errorf("state for resource is not serializable to json: %w", err)
+	}
+
+	stateProtoAny, err := component.ProtoAny(r.State())
```

Comment:
@mitchellh Is this what you were thinking should be saved here? There's a `.proto` method we could use here instead, but I don't think the `Framework_ResourceState` that it returns is what we care about here.